### DISCUSSION TITLE
How do I type `set` to allow setting nested objects without `type is missing properties` error?

### DISCUSSION THREAD

**Discussion by b-novikov-ipersonality** [pmndrs/zustand#945] [Q&A] [answered]
Hi guys. 

[Codesandbox](https://codesandbox.io/s/red-http-jj8kbx?file=/src/typeerror.ts)

As I understand zustand makes deep merge so I can change state of specific property without returning the whole new state. 

I have some code:

```javascript
const createModalSlice = (
  set: SetState<ModalStore>,
  _get: GetState<ModalStore>
): ModalStore => ({
  isModalOpen: false,
  modal: {
    title: '',
    text: '',
    onClose: () => set({ isModalOpen: false, modal: null }),
    onOpen: (title: string, text: string) =>
      set({ isModalOpen: true, modal: { title, text } }), // typescript error here
  },
});
```

But I get typescript error `Type '{ title: string; text: string; }' is missing the following properties from type 'ModalType': onClose, onOpen`

Type of `modal`:
```javascript
type Modal = null | {
  title: string;
  text: string;
  onAction?: () => void;
  onClose: () => void;
  onOpen: (title: string, text: string) => void;
};
```
As I understand the problem is with `set` typing. How do I type it properly?

**Comment by dai-shi** [comment, marked as answer]
No, zustand does shallow merge. So, types are correct.
FYI: https://github.com/pmndrs/zustand/wiki/Updating-nested-state-object-values

**Reply by b-novikov-ipersonality** [reply]
So if I set `modal` to be equal to `{ title, text }` all other properties would be "deleted"?

**Reply by dai-shi** [reply]
Yes, exactly.

You want to do this:
```js
    onOpen: (title: string, text: string) =>
      set((state) => ({ isModalOpen: true, modal: { ...state.modal, title, text } })),
```
